Merged with my previous repo.

JustAnotherDevv opened the pull request:
Changes after last feedback:

- Support for paginated results when requesting tickets or user NFTs
- Support for minting more than 250 NFTs by creating tickets dynamically based on currently existing tickets, XRP balance etc.
- Verification for user based on event minter wallet address and event id that's stored in metadata as taxon.
- Attendees lookup based on participants list stored in local participants.json file.
- Better documentation for class.
- All expressJS routes are also commented now and additional docs for those is provided in readme.md.
- Other small improvements were made

Based on the bounty description and previous comments I believe that it should be sufficient to complete 1, 2 and 3 milestones. If there is need for any changes before submission could be officially accepted please let me know.

In upcoming days/weeks I am planning on adding extra voluntary additions like the hosted docs and video demo, additional features.

JustAnotherDevv committed:
Added files from my local repo.

JustAnotherDevv committed:
Small bug fixes for checking whether user requesting claim is on the participants list already.

JustAnotherDevv committed:
Update README.md

JustAnotherDevv committed:
Update README.md

jonathanlei commented:
> @jonathanlei Just wondering if there is still something I have to do to push review process by your team forward. Do I have to fill out the bounty form again?

hey! We are still reviewing the bounty with @JST5000, thanks for your patience!

JST5000 commented:
I'll review this tomorrow :)

JST5000 commented on README.md:
```diff
@@ -1,2 +1,113 @@
+
+## Project information
+
+Project was firstly developed during [XRPL NFT hackathon.](https://devpost.com/software/xrp-nft-attendence)
```
Could you include a 1-2 high level description of it at the top? (You can maybe borrow the preview one of "Attendify is an API to allow event organizers to mint and distribute NFTs on the XRP Ledger to track and reward attendance at their events."

That way when they get to the features, they have a big picture to attach the individual use cases to :)

JustAnotherDevv replied:
Added description, also the docs are now [hosted online just like you requested](https://justanotherdevv.github.io/AttendifyXRPL/Attendify.html).
I have requested review of changes.

JustAnotherDevv committed:
changed docs to host them on github pages.

JustAnotherDevv committed:
Merge branch 'master' of https://github.com/JustAnotherDevv/AttendifyXRPL

JustAnotherDevv committed:
Update README.md

JustAnotherDevv committed:
Update README.md

JustAnotherDevv committed:
Update README.md

JustAnotherDevv committed:
Update README.md

JST5000 commented:
Started trying to use this locally to see how it works in practice, some initial feedback from that:
1. Could you add examples of the requests to the README? (or dummy text to copy with the paramaters set up - trying to go back & forth to create the URL I needed to call the endpoint was a bit annoying)
Ex. http://localhost:4000/api/claim?walletAddress=rpHJL8bRW81shYm3LwwsXovABps6SM51RS&type=2&minter=rMBU1wwAJZZz36YBNxug7YYHKeC5k3vqfG&eventId=0 or http://localhost:4000/api/claim?walletAddress=&type=&minter=&eventId=
2. The event id returned from `api/mint` is called `id`, which is a bit confusing. Instead I'd recommend naming it `eventId` to be in line with the parameter in `api/claim`

Question: Is the `minter` returned from `api/claim` the one I specified in `.env`? 

(I'll keep testing this tomorrow as well)

JST5000 commented on index.js:
```diff
@@ -0,0 +1,247 @@
+      const { walletAddress, tokenCount, url, title, desc, loc } =
+        await req.query;
+      if (
+        walletAddress.length == 0 ||
+        tokenCount.length == 0 ||
+        url.length == 0 ||
+        title.length == 0 ||
+        walletAddress.length == 0 ||
+        desc.length == 0 ||
+        loc.length == 0
```
When I forgot a paramater or mistyped one, these checks error because the value is actually undefined instead of having a length of 0. (At least on my machine) so they raise the opaque error: “TypeError: Cannot read properties of undefined (reading 'length')” instead of one which tells me which paramaters I messed up.

JustAnotherDevv replied:
That's a good point, in my case I had to check for empty strings but I'll add check for null values too.

Also regarding comment above: `api/claim` currently returns just status of claim and the list with NFTs for selected event that aren't claimed yet but I suppose you were asking for `minter` parameter that's passed to this endpoint. In current implementation usually it's wallet address for account from `.env` but with modified implementation this could support multiple minter accounts.

JST5000 commented on index.js:
```diff
@@ -0,0 +1,247 @@
+      console.log(metadata);
+
+      return res.send({
+        result: await AttendifyLib.batchMint(
```
While batchMint does now use Tickets, it seems to still be doing one mint per ledger as opposed to leveraging the tickets to do a couple at the same time (Screenshot attached) - For small numbers it's not a big deal, but it could lead to long wait times as you start minting 1000+ NFTs. 
![image](https://user-images.githubusercontent.com/19694186/214731264-1e22ebc0-17dc-4c69-8b3a-1fd86e214930.png)

JST5000 replied:
(Here I minted 5 NFTs using api/mint, and they were actually minted about once per ledger)

JustAnotherDevv replied:
That was caused by old code that used  `submitAndWait` instead of `submit`. I have already changed it and from what I've seen in explorer multiple NFTs are minted at the same time now.

JST5000 commented on README.md:
```diff
@@ -1,2 +1,109 @@
+
+**@param {string} walletAddress** - Wallet address from user requesting claim
+
+**@param {integer} tokenCount** - Amount of NFTs that should be create for the event
```
nit: 
```suggestion
**@param {integer} tokenCount** - Amount of NFTs that should be created for the event
```

JST5000 commented on README.md:
```diff
@@ -1,2 +1,109 @@
+
+**@param {string} walletAddress** - Wallet address from user requesting verification
+
+**@param {string} signature** - Signature from requesting user
```
What do they have to sign? I'm a bit unclear what I'm supposed to put into this parameter. (Is it a fake transaction like XUMM does?)

JST5000 replied:
(If it can be any transaction I'd recommend expanding this description to include an example of what you could sign - Ex. An `AccountSet` transaction with `SetFlag` set to 0 (no flags) is a common useless operation you can have people sign)

JustAnotherDevv replied:
Yeah, it's just a signature to prove ownership of `walletAddress`. In integration with UI this could be done using XUMM, currently for the purpose of testing I am firstly getting user wallet `const myWallet = xrpl.Wallet.fromSeed(walletForSignatureVerification.seed);` and then using function from xrpl.js to create signature 
`let my_seq = 21404872;
    const txJSON = {
      Account: myWallet.address,
      TransactionType: "Payment",
      Destination: "rf1BiGeXwwQoi8Z2ueFYTEXSwuJYfV2Jpn",
      Amount: "13000000",
      Flags: 2147483648,
      LastLedgerSequence: 7835923,
      Fee: "13",
      Sequence: my_seq,
    };
    const signature = await myWallet.sign(txJSON);`

JST5000 commented on README.md:
```diff
@@ -1,2 +1,109 @@
+
+    -`transferred` - Indicates that new sell offer for NFT related to selected event was created successfully and details were sent to user
+
+⚫ `GET /api/verifyOwnership` - Verifies whether or not user owns NFT with provided id for particular user.
```
When I try this with: http://localhost:4000/api/verifyOwnership?walletAddress=rwbq1nusPZmRoQCSgwjNXGch8pJyhBJqEs&signature=1200037321ED62F3992FE1E78FF2FF4A2A908F6B154B646298435A0A0018929A440BDA976A4E744068BA83C4525F7122DEB627DB0CF5EBE86A254E226D763044C69EF27C412F3C67C426159631C5DCA27C863F5D76830D02140D2C53F9C8BC6BE0724D8904C1FA0B8114695010606A0E859D8B4E4D4F41F57E5C0BD389AD&minter=rMBU1wwAJZZz36YBNxug7YYHKeC5k3vqfG&eventId=0

(which is a wallet that has just successfully done the `api/claim` call, with a signed `AccountSet` transaction that does nothing) 


I get `false` returned - I'm a bit confused, do I need to do more after `api/claim` in order for that account to actually own the NFT, or am I maybe doing the signature wrong? 

Can you provide an example of successfully calling this function so I can understand what I'm doing wrong?

JustAnotherDevv replied:
If you're unsure about correct way to call some of the functions I would recommend checking the way it's done in `test.js` file, you can also run it using command `npm run test` to check the output and some of the logs.

In current implementation `/api/attendees` endpoint returns that user is one of attendees immediately after claim process was started. Meaning that the user requested claim and new NFT sell offer was created for him but it still has to be accepted by the user manually. 

On the other hand `/api/verifyOwnership?walletAddress` endpoint performs actual check using ledger data to see if user has NFT from selected event. That's why it returns `false` unless you accepted the NFT offer that was returned from `/api/claim` enpoint.

I hope that answers your question, this was conscious design decision I made to support different use cases but I am open to changes if you consider that to be important.

JustAnotherDevv committed:
Added API request examples

JustAnotherDevv committed:
Update README.md

Co-authored-by: Jackson Mills <aim4math@gmail.com>

JustAnotherDevv committed:
Fixed typos and added testing instructions

JustAnotherDevv committed:
Added explanation for signatures

JustAnotherDevv committed:
Added complete param checks for API routes | Batch minting now supports minting few NFTs at the same time thanks to tickets | Error messages were updated

JustAnotherDevv committed:
Updated example API request

JustAnotherDevv committed:
Update README.md

JustAnotherDevv committed:
Update README.md

JustAnotherDevv commented:
I have completed required changes and recorded a short video showcasing how to use Attendify - https://youtu.be/YLdJcGXUpjI
Examples of requests and updated information were added to readme file.

JustAnotherDevv committed:
Updated jsdoc docs

JST5000 commented:
Thanks for making those changes! I'll review this again tomorrow :)

JST5000 commented:
When you do `npm i` there are 12 dependencies with "High security warnings" - it's worth looking into them to see if you can replace them with secure replacements, or limit the impact of those vulnerabilities.

JST5000 commented:
When I run `npm run test` I get a couple errors in the console: 
```
  Testing typical user flow
{
  publicKey: 'ED7E0C656AA135DB06915D9DB6EABCCD56E88271DEDD3B5BFC358D3D84387504C5',
  privateKey: 'ED95B30AAA6CCEC20AE5C898D4A6C10459B351D233C44C3C69B6E29EE32760F504',
  classicAddress: 'ra2QmiPCKPed5Bd5tP3HiN6UdQSPgcvZFn',
  seed: 'sEdSTNDDxvYHfwdg6AuoTwG9XnKRBSJ'
}
    ✔ Getting new wallet account (6364ms)
IPFS error  TypeError: RequestInit: duplex option is required when sending a body.
    at Object.fetch (node:internal/deps/undici/undici:14062:11)
    at async Client.fetch (/Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/node_modules/ipfs-utils/src/http.js:132:22)
    at async addAll (file:///Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/node_modules/ipfs-http-client/src/add-all.js:36:17)
    at async last (file:///Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/node_modules/it-last/dist/src/index.js:7:22)
    at async Object.add (file:///Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/node_modules/ipfs-http-client/src/add.js:22:14)
    at async postToIPFS (/Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/utils.js:51:20)
    at async /Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/index.js:81:26
TypeError: RequestInit: duplex option is required when sending a body.
    at Object.fetch (node:internal/deps/undici/undici:14062:11)
    at async Client.fetch (/Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/node_modules/ipfs-utils/src/http.js:132:22)
    at async addAll (file:///Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/node_modules/ipfs-http-client/src/add-all.js:36:17)
    at async last (file:///Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/node_modules/it-last/dist/src/index.js:7:22)
    at async Object.add (file:///Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/node_modules/ipfs-http-client/src/add.js:22:14)
    at async postToIPFS (/Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/utils.js:51:20)
    at async /Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/index.js:81:26
marker,  undefined
Max tickets 250
marker,  undefined
minting  1 / 3  NFTs
TypeError [ERR_INVALID_ARG_TYPE]: The first argument must be of type string or an instance of Buffer, ArrayBuffer, or Array or an Array-like Object. Received an instance of TypeError
    at new NodeError (node:internal/errors:400:5)
    at Function.from (node:buffer:332:9)
    at Object.convertStringToHex (/Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/node_modules/xrpl/dist/npm/utils/stringConversion.js:5:19)
    at Attendify.batchMint (/Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/attendify.js:303:23)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/index.js:86:19 {
  code: 'ERR_INVALID_ARG_TYPE'
}
Error: TypeError [ERR_INVALID_ARG_TYPE]: The first argument must be of type string or an instance of Buffer, ArrayBuffer, or Array or an Array-like Object. Received an instance of TypeError
    at Attendify.batchMint (/Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/attendify.js:335:13)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/index.js:86:19
undefined
    1) Minting NFTs for new event
    2) Checking if it's possible to claim NFT for event
    3) Claiming offer for NFT from event
    4) Looking up if the test user is on the attendees list for test event
minter  rMBU1wwAJZZz36YBNxug7YYHKeC5k3vqfG
signature  12000022800000002401469CC8201B00779113614000000000C65D4068400000000000000D7321ED57D41105FC480545763677D2100C8949324A97811FE5CB45594B5E73991BBF927440625FEB2E8DEA86736368DE716C0223586C97C44519B45500653F7C505DA9C24F3F8789BBEE917871DC862DB826ACE7B64AB485BADB0C46776E931EA0EC55840381143CBB7B1E0212681492733BDA77986A6A7C4C2B4A83144B4E9C06F24296074F7BC48F92A97916C6DC5EA9
    5) Verifying ownership of NFT


  1 passing (12s)
  5 failing

  1) Testing typical user flow
       Minting NFTs for new event:

      AssertionError: expected 500 to equal 200
      + expected - actual

      -500
      +200
      
      at /Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/test.js:54:33
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

  2) Testing typical user flow
       Checking if it's possible to claim NFT for event:
     TypeError: Cannot read properties of undefined (reading 'eventId')
      at Context.<anonymous> (test.js:79:99)
      at process.processImmediate (node:internal/timers:471:21)

  3) Testing typical user flow
       Claiming offer for NFT from event:
     TypeError: Cannot read properties of undefined (reading 'eventId')
      at Context.<anonymous> (test.js:93:99)
      at process.processImmediate (node:internal/timers:471:21)

  4) Testing typical user flow
       Looking up if the test user is on the attendees list for test event:
     TypeError: Cannot read properties of undefined (reading 'eventId')
      at Context.<anonymous> (test.js:107:65)
      at process.processImmediate (node:internal/timers:471:21)

  5) Testing typical user flow
       Verifying ownership of NFT:
     TypeError: Cannot read properties of undefined (reading 'eventId')
      at Context.<anonymous> (test.js:134:187)
 ```

JST5000 commented:
Similar error occurs when using the first example you added to the README: 

INPUT: http://localhost:4000/api/mint?walletAddress=rpHJL8bRW81shYm3LwwsXovABps6SM51RS&tokenCount=5&url=ipfs://QmQDDD1cNgnyhPC4pBLZKhVeu12oyfCAJoWr1Qc1QgbkPN&title=test_title&desc=test_description&loc=Warsaw

OUTPUT: 
{"statusText":"Error: TypeError [ERR_INVALID_ARG_TYPE]: The first argument must be of type string or an instance of Buffer, ArrayBuffer, or Array or an Array-like Object. Received an instance of TypeError"}

JST5000 commented on package.json:
```diff
@@ -0,0 +1,33 @@
+{
+  "dependencies": {
+    "axios": "^1.2.2",
```
It seems `mocha` should be a dev dependency (Otherwise when you run `npm i` followed by `npm run test` it error's out saying it doesn't know what `mocha` is)

JustAnotherDevv commented:
- I somehow missed the aprt with adding `mocha` to devDependencies because it's installed globally on my PC.
- It looks like the vulnerabilities were in the unused babel dependencies from packages that I have used previously but are replaced by different code now. I still have to test a couple of small thins but I think it fully solved the issue. Also I have removed few other unused dependencies.
- Regarding the error that you're getting now - it seems like it's somehow related to IPFS 
> IPFS error  TypeError: RequestInit: duplex option is required when sending a body.
There are 2 different environmental variables that are crucial for this part of the code to work correctly: `INFURA_ID` and `INFURA_SECRET`. Please make sure to double check whether you have added them to your project from INFURA dashboard. These key names are `API Key Secret` and `PROJECT ID`, here's a [guide showcasing how you can quickly get these keys](https://docs.infura.io/infura/getting-started). I will also add the IPFS setup instructions to readme file.

I think that fully solves the problems related to setup, if you still have any issues with that maybe we can somehow schedule a quick call where I will go through the IPFS related stuff. I will push the updated code soon once I finish review of changes.

JustAnotherDevv committed:
Removed unused npm packages & added IPFS setup instruction.

JustAnotherDevv committed:
Removed unused npm packages & added IPFS setup instruction.

JST5000 commented:
Please also add `nodemon` to the dev dependencies :) 

Re: Infura setup, I'm using the same keys as I used to originally play with Attendify which worked before, so they should be working fine. (I double checked the file is configured the same)

It seems that the error I'm encountering has to do with the use of the function `btoa`, which appears to be deprecated & isn't able to be interpretted on my machine. (IPFS is giving the error for some reason, but when I hover btoa it says it shouldn't be used for new code - screenshot attached)

```
XRPL Attendify server listening on port 4000
IPFS error  ReferenceError: **btoa is not defined**
    at postToIPFS (/Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/utils.js:44:7)
    at async /Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/index.js:81:26
ReferenceError: btoa is not defined
    at postToIPFS (/Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/utils.js:44:7)
    at async /Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/index.js:81:26
marker,  undefined
Max tickets 236
marker,  undefined
minting  1 / 5  NFTs
TypeError [ERR_INVALID_ARG_TYPE]: The first argument must be of type string or an instance of Buffer, ArrayBuffer, or Array or an Array-like Object. Received an instance of ReferenceError
    at new NodeError (internal/errors.js:322:7)
    at Function.from (buffer.js:334:9)
    at Object.convertStringToHex (/Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/node_modules/xrpl/dist/npm/utils/stringConversion.js:5:19)
    at Attendify.batchMint (/Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/attendify.js:303:23)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async /Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/index.js:86:19 {
  code: 'ERR_INVALID_ARG_TYPE'
}
Error: TypeError [ERR_INVALID_ARG_TYPE]: The first argument must be of type string or an instance of Buffer, ArrayBuffer, or Array or an Array-like Object. Received an instance of ReferenceError
    at Attendify.batchMint (/Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/attendify.js:335:13)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async /Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/index.js:86:19
```
<img width="557" alt="image" src="https://user-images.githubusercontent.com/19694186/216191953-d9c7e87a-662e-4bc1-a226-9414e572333b.png">

JST5000 commented:
Ah sorry, updating my node version to 16 from 14 fixes it (Although `btoa` is still deprecated)

JST5000 commented:
I just realized there's a potential issue with allowing any signed data to validate ownership over the account - that doesn't necessarily prove ownership because someone could just take a transaction which was signed by the legitimate account owner instead of signing it themselves. (When you submit a transaction to the ledger, you have to sign it, and that data is transmitted over the internet both in submitting a transaction and receiving a response from rippled in the `tx_blob` field) 

So, in order to actually verify that the only someone who knows the secret to an account can get a "true" response from the `verifyOwnership` you may have to have them sign a transaction that wouldn't be used normally on mainnet. Maybe including a number that you provide to ensure they signed it just now. (That way someone couldn't just use an intercepted transaction from the person who actually owns the account)

JustAnotherDevv committed:
Added unique IDs for ownership verification

JustAnotherDevv commented:
@JST5000 I'm happy to see that it finally worked for you, already updated btoa to different implementation that's not deprecated. 

Regarding signatures: now I am using new endpoint `startVerification` to generate random ID for each user, storing it temporarily and then user requesting verification has to include this ID in memo. Memo ID must match expected ID that was generated previously. The chance of submitting correct ID by attacker that's spamming random strings of characters is so low it could be considered impossible in practice.

JustAnotherDevv committed:
Added checks to ensure that MEMO_ID isn't null

JustAnotherDevv committed:
Updated documentation

JST5000 commented:
I'll review again tomorrow - nice solve :)

JST5000 commented on README.md:
```diff
@@ -1,2 +1,127 @@
+
+**@param {string} walletAddress** - Wallet address from user requesting verification
+
+Example request `http://localhost:4000/api/startVerification?walletAddress=raY33uxEbZFg7YS1ofFRioeENLsVdCgpC5`
```
For `startVerification` can you also provide an example transaction that can be signed to show how to add it to the memo? (There are a lot of different ways of putting it into the memo - should it be hex encoded, normal string, etc.)

JST5000 commented on test.js:
```diff
@@ -0,0 +1,170 @@
+    let my_seq = 21404872;
+    const txJSON = {
+      Account: myWallet.address,
+      TransactionType: "Payment",
```
As the example, instead of a payment, I'd recommend having them do a dummy `AccountSet` transaction because otherwise people may accidentally create a valid transaction which can send funds if someone intercepts it and actually executes it.

JST5000 commented on README.md:
```diff
@@ -1,2 +1,127 @@
+**@param {string} minter** - The minter address of the event
+
+**@param {string} eventId** - The ID of the event
+
```
Could you add examples of the return types? (That'll be useful for other programs trying to integrate with this)

JST5000 commented on README.md:
```diff
@@ -1,2 +1,127 @@
+
+**@param {string} eventId** - The ID of the event
+
+Example request `http://localhost:4000/api/verifyOwnership?walletAddress=raY33uxEbZFg7YS1ofFRioeENLsVdCgpC5&signature=12000022800000002401469CC8201B00779113614000000000C65D4068400000000000000D7321ED57D41105FC480545763677D2100C8949324A97811FE5CB45594B5E73991BBF927440625FEB2E8DEA86736368DE716C0223586C97C44519B45500653F7C505DA9C24F3F8789BBEE917871DC862DB826ACE7B64AB485BADB0C46776E931EA0EC55840381143 ...
```
Also, it's quite hard to debug what is incorrect when using `verifyAccount` - is it that my issuer doesn't exist? That my signed transaction isn't correct? Did I forget to include the `startVerification` memo? Did I accidentally provide the wrong number of parameters? (I spent a while trying to get this to work with a transaction I signed locally, but couldn't figure it out before the end of my day unfortunately - I'd recommend adding more clear errors to help people in the future debug where their integration is running into trouble :) )

Will continue testing on Monday :)

JustAnotherDevv committed:
Added examples to readme and more detailed error messages.

JustAnotherDevv committed:
New negative test cases for NFT ownership verification.

JustAnotherDevv committed:
Added example of running local.js

JustAnotherDevv commented:
I've pushed some tweaks that should make it easier for you to debug & test the code

- Changes that you requested are in readme file already
- Errors should provide now more descriptive information
- Negative test cases showcasing incorrect use of parameters were added
- I pushed to git the `local.js` that can be used for faster manual creation of signatures

Run this file locally with 2 arguments just like the example below to generate signed tx blob

- The first argument is seed for the user wallet
- The second argument should be verification ID returned from `api/startVerification` endpoint
  `node .\local.js sEdVMJSLjuTAjaSeeZ6TEkpUWuTS83j PPrnp3O4FI3Z5VCnMvvsMwzjMCmrtZnUNqWQEqkGXIuGJ7xzpCdhUql3X9SwlP1d1unotJX8FHCMJTgbYUQfa7b1nSVb7FY9wSrWH3NxQIF9py1MXfOGGiZyjM6S4WuxhxK1Ru499u9jhnZ3vhifyRPVYFytDD2YUsjqG7F3ojWkVC4VoOEHP4uuztfb2k0ZueCIP1tfbgoAXEuU5DxoSLAWx9kP3b7nKuOqV6lvlbKWGzCI5xE1YG5ExTkowk0M`

JST5000 commented on attendify.js:
```diff
@@ -0,0 +1,438 @@
+      const EXPECTED_MEMO_ID = this.signatureMap.get(walletAddress);
+      if (!EXPECTED_MEMO_ID)
+        throw new Error(
+          `Wallet address '${walletAddress}' don't have the verification ID generated for it yet. Please start verification process by obtaining a verification ID before performing ownership verification check.`
```
nit: 
```suggestion
          `Wallet address '${walletAddress}' doesn't have the verification ID generated for it yet. Please start the verification process by calling `startVerification` and including the returned value as part of the memo in your signed transaction when calling this function.`
```

JST5000 commented on attendify.js:
```diff
@@ -0,0 +1,438 @@
+      );
+      if (accountNfts.length == 0) return false;
+      for (let i = 0; i != accountNfts.length; i++) {
+        if (accountNfts[i].Issuer == minter) {
```
When you call `verifyOwnership` with the wrong issuer, this errors out in the following way. It'd be nice to give a nice error message like "That account doesn't exist, did you accidentally copy the 'owner' instead of the 'minter'?"

(The object printed out is `accountNfts[i]` right before `accountNfts[i].Issuer` throws the error and comes from a print statement I added, not in the native code) 

```
 data: {
    error: 'actNotFound',
    error_code: 19,
    error_message: 'Account not found.',
    id: 0,
    request: {
      account: 'rN8gWXhMyec54tpdKeE9ZFd717Qf3ViFoF',
      id: 0,
      method: 'account_nfts'
    },
    status: 'error',
    type: 'response'
  }
}
undefined
TypeError: Cannot read properties of undefined (reading 'Issuer')
    at Attendify.verifyOwnership (/Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/attendify.js:376:28)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async /Users/jacksonmills/Documents/bounty_submissions/AttendifyXRPL/index.js:230:37
```

JST5000 replied:
This error is also caused by accidentally using a valid account from the wrong network - so the signature works fine, but this lookup fails.

JST5000 commented on attendify.js:
```diff
@@ -0,0 +1,438 @@
+          `Memo '${TX_MEMO}' from signature does not match expected ID for the provided wallet address '${walletAddress}'.`
+        );
+      // Getting user NFTs and checking whether any NFT was issued by minter address
+      const accountNfts = await await this.getBatchNFTokens(
```
You're calling await twice
```suggestion
      const accountNfts = await this.getBatchNFTokens(
```

JST5000 commented on attendify.js:
```diff
@@ -0,0 +1,438 @@
+      });
+      let accountNfts = nfts.result.account_nfts;
+      //console.log("Found ", accountNfts.length, " NFTs in account ", address);
+      for (;;) {
```
Should this be a while loop instead?

JST5000 commented on attendify.js:
```diff
@@ -0,0 +1,438 @@
+   * @returns {object} - The metadata of the sell offer for a given NFT from selected event
+   * @throws {Error} - If any of the required parameters are missing or if there is an issue creating the sell offer
+   */
+  async createSellOfferForClaim(buyer, minterSeed, TokenID) {
```
I think `claim` shouldn't allow you to claim an NFT from an account which doesn't exist on ledger. (I accidentally made an account on Testnet instead of Devnet, the default network, and this still sold to that account even though it wasn't a valid account on ledger)

JST5000 commented:
Almost got things working for the verifyAccountOwnership, but only realized at the end I need to claim the NFTokenOffer, so will pick this back up tomorrow to figure out how to do that locally in order to make sure everything works :)

JST5000 commented:
I'm struggling to find a good way to claim the NFTs that have offers out - do you know if there's a way to do it with Xumm or something like that? (Or a small script?) I can take another crack tomorrow but I'm not sure how event organizers would get attendees to actually claim the NFTs after minting them currently.

JustAnotherDevv committed:
Update attendify.js

Co-authored-by: Jackson Mills <aim4math@gmail.com>

JustAnotherDevv committed:
Update typo in  attendify.js

Co-authored-by: Jackson Mills <aim4math@gmail.com>

JustAnotherDevv committed:
Added checks to see if account exists on the ledger and script with example of accepting NFT offer.

JustAnotherDevv committed:
fixed typo

JustAnotherDevv commented:
- I've added `checkIfAccountExists` method in the Attendify that's called a few times and checks whether or not wallet address that was given actually exists on the selected network. That should check problems that you mentioned.
- Changed 2 loops from `for(;;)` to `white (true)` after your request although this has no effect on the code.
- Previous `local.js` now is now named `signature.js` and can be used in the same way. New file `acceptOffer.js` contain example of possible implementation for the end use accepting the NFT claim offer. Example usage for your test:
`node .\acceptOffer.js sEdVMJSLjuTAjaSeeZ6TEkpUWuTS83j 80517CC6087108B777710DFD0C48B6CB66A43947A96CA4C4B145574E27E9749A`
- The first argument is seed for the user wallet

 - The second argument should be NFT offer IF returned from `api/mint` endpoint

JustAnotherDevv committed:
Chaned readme.

JST5000 commented on acceptOffer.js:
```diff
@@ -0,0 +1,89 @@
+/**
+ * Run this file locally with 2 arguments just like the example below to generate signed tx blob
+ * - The first argument is seed for the user wallet
+ * - The second argument should be NFT offer ID returned from `api/mint` endpoint
```
```suggestion
 * - The second argument should be NFT offer ID returned from `api/claim` endpoint
```

JST5000 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+
+    -`success` - If `onlyCheckStatus` parameter was true it indicated that current parameters for selected event were sent
+
+    -`transferred` - Indicates that new sell offer for NFT related to selected event was created successfully and details were sent to user
```
Would it be possible to simplify the return type when it is successfully transferred to just show the NFTokenOffer the user gets instead of also showing the other NFTokens that are available? (It is a lot of information that the end user can't use, and is easy to get confused about which value to copy for the next step in the process)

JST5000 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+
+    -`404` - Event with specified ID does not exist
+
+    -`claimed` - The NFT for specified event was already claimed by this user
```
When the NFTokens are all claimed, I don't think the end user should see a list of accounts which received those NFTs already. If I was an attendee who didn't make the cut, I wouldn't need to know who else did get there before I did.

JST5000 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+
+  `node .\acceptOffer.js sEdVMJSLjuTAjaSeeZ6TEkpUWuTS83j 80517CC6087108B777710DFD0C48B6CB66A43947A96CA4C4B145574E27E9749A`
+
+Example request `http://localhost:4000/api/verifyOwnership?walletAddress=raY33uxEbZFg7YS1ofFRioeENLsVdCgpC5&signature=120003240000000520210000000568400000000000000C722103AB40A0490F9B7ED8DF29D246BF2D6269820A0EE7742ACDD457BEA7C7D0931EDB7321ED57D41105FC480545763677D2100C8949324A97811FE5CB45594B5E73991BBF927440DC49BE4996DFF6CF55D61C7689CF212CCDE1A8F183A30516CDBE8117947C3CD561DAFCFC970504A564C7C9E948F ...
```
This request is for later on
```suggestion
```

JST5000 commented:
Sweet, with the `acceptOffer.js` script and `signature.js` script I'm able to completely use your API :)

JustAnotherDevv committed:
Update acceptOffer.js

Co-authored-by: Jackson Mills <aim4math@gmail.com>

JustAnotherDevv committed:
Update README.md

Co-authored-by: Jackson Mills <aim4math@gmail.com>

JustAnotherDevv committed:
Removed unnecessary information from claim endpoint return

JustAnotherDevv committed:
Checking ability to claim now returns amount of claimable NFTs

JustAnotherDevv commented:
I've commited requested changes. Now only if user wants to check if he's eligible for claim the `/api/claim` endpoint returns amount of NFTs that are still available. Documentation was also regenerated. 

It's great that you want to finally approve the changes, btw do I have to manually add the other person for review too? It somehow unassigned other reviewer when I requested re-review from you. Github UI is kinda confusing in this case.

jonathanlei reviewed (approved):
LGTM, thanks for all of your hard work! Once @JST5000 take a final look we will merge and start the payout process for the first 3 milestones.

JST5000 reviewed (approved):
LGTM after fixing `npm run test` :)

JustAnotherDevv committed:
Updated tests and added new video demo

JustAnotherDevv commented:
The old test step for `verifyOwnership` had other account that already had accepted NFT offer from the minter account. In my updated test now it's automatically accepting offer from `claim` endpoint each time. I have also created new video with updated version of the API, it's in the readme already.

JST5000 commented:
Sweet, thanks for doing that!

JST5000 commented:
Merged! This satisfies Milestones 1-3 of the POAP bounty, you should submit to claim that bounty here: [https://xrpl.typeform.com/xrpl-bounties](https://xrpl.typeform.com/xrpl-bounties?utm_source=xxxxx&utm_medium=xxxxx&utm_campaign=xxxxx&utm_term=xxxxx&utm_content=xxxxx)
